### PR TITLE
update docs

### PR DESCRIPTION
**Pull request by metaflow**
- move "user_doc" to "README" as it's publicly facing page visible when one opens repo
- move developer docs to "development"
- update docs to reflect current buildkite setup (not completely)
- grammar

new diagramm:
d

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,39 +1,66 @@
 # Overview
 
-This repository contains the configuration files for the pre-merge checks for the LLVM project. This github project contains the documentation and the server configuration cluster of build machines that are used to check all incoming commits to the LLVM project.
+The *pre-merge checks* for the [LLVM project](http://llvm.org/) are a [continuous integration (CI)](https://en.wikipedia.org/wiki/Continuous_integration) workflow. The workflow checks the patches the developers upload to the [LLVM Phabricator](https://reviews.llvm.org) instance. *Phabricator* is the code review tool in the LLVM project. The workflow checks the patches before a user merges them the master branch - thus the term *pre-merge testing*. When a user uploads a patch to the LLVM Phabricator, Phabricator triggers the checks and then displays the results.
```
Comment (ChristianKuehnel) [inline README.md:3]: add "to": them **to** the master branch/

Comment (ChristianKuehnel) [inline README.md:3]: Nit: Maybe wrap text at 80 or 120 characters for easier reviews and diffing.

File: README.md [hunk 2]
```diff
@@ -1,39 +1,66 @@
...
+
+* You set a git hash as `sourceControlBaseRevision` in Phabricator which is available on the Github repository,
+* **or** you define the dependencies of your patch in Phabricator, 
+* **or** your patch can be applied to the master branch.
+
+Only then can the build server apply the patch locally and run the builds and tests.
+
+# Accessing results on Phabricator 
+
+TODO: move closer to top and explain what different links mean currently. Update screenshots.
```
Comment (ChristianKuehnel) [inline README.md:40]: TODO in the text

File: README.md [hunk 3]
```diff
@@ -1,39 +1,66 @@
...
+
+Phabricator will automatically trigger a build for every new patch you upload or modify.
+Phabricator shows the build results at the top of the entry:
+![build status](docs/images/diff_detail.png)
+
+CI will compile and run tests, run clang-format and [clang-tidy](docs/clang_tidy.md) on lines changed.
```
Comment (ChristianKuehnel) [inline README.md:46]: add "the": **The** CI...

File: README.md [hunk 4]
```diff
@@ -1,39 +1,66 @@
...
+
+# Contributing
+
+We're happy to get help on improving the infrastructure and workflows!
+
+Please check [contibuting](docs/contributing.md) first.
+[Development](docs/development.md) gives an overview how different parts interact together.
+[Playbooks](docs/playbooks.md) shows concrete examples how to, for example, build and run agents locally.
+
+If you have any questions please contact [Christian Kühnel](mailto:kuhnel@google.com) or [Mikhail Goncharov](mailto:goncahrov@google.com).
```
Comment (ChristianKuehnel) [inline README.md:59]: Maybe remove me from the list?

File: docs/development.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
+- [Overview](#overview)
+- [Phabricator integration](#phabricator-integration)
+- [Buildkite pipelines](#buildkite-pipelines)
+- [Life of pre-merge check](#life-of-pre-merge-check)
+- [Enabled projects and project detection](#enabled-projects-and-project-detection)
+- [Agent machines](#agent-machines)
+- [Compilation caching](#compilation-caching)
+- [Buildkite monitoring](#buildkite-monitoring)
+
+<small><i><a href='http://ecotrust-canada.github.io/markdown-toc/'>Table of contents generated with markdown-toc</a></i></small>
```
Comment (ChristianKuehnel) [inline docs/development.md:10]: Where does this come from?

File: docs/development.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+# Overview
+
+* [Buildkite](https://buildkite.com/llvm-project) orchestrates each build.
+* multiple Linux and windows agents connected to Buildkite. Agents are run at Google Cloud Platform.
+* [small proxy service](/phabricator-proxy) that takes build requests from [reviews.llvm.org](http://reviews.llvm.org) and converts them into Buildkite build request. Buildkite job sends build results directly to Phabricator.
```
Comment (ChristianKuehnel) [inline docs/development.md:16]: I would rephrase a bit:
* [small proxy service](/phabricator-proxy) receives build requests from [reviews.llvm.org](http://reviews.llvm.org) and triggers Buildkite build requests. The Buildkite job sends build results directly to Phabricator.

File: docs/development.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+* every review creates a new branch in [fork of llvm-project](https://github.com/llvm-premerge-tests/llvm-project).
```
Comment (ChristianKuehnel) [inline docs/development.md:17]: s/review/uploaded diff/

File: docs/playbooks.md [hunk 8]
```diff
@@ -1,5 +1,20 @@
...
+    + [Buildkite](#buildkite)
+    + [Jenkins](#jenkins)
+  * [Testing scripts locally](#testing-scripts-locally)
+  * [Custom environment variables](#custom-environment-variables)
+  * [Testing changes before merging](#testing-changes-before-merging)
+- [Phabricator integration](#phabricator-integration)
+  * [Herald](#herald)
+  * [Harbormaster](#harbormaster)
+  
+<small><i><a href='http://ecotrust-canada.github.io/markdown-toc/'>Table of contents generated with markdown-toc</a></i></small>
```
Comment (ChristianKuehnel) [inline docs/playbooks.md:15]: Same: Where does this come from?

File: docs/development.md [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+# Phabricator integration
+
+* [Harbormaster build plan](https://reviews.llvm.org/harbormaster/plan/5On the Phabricator side these things were configured:/)
+* Herald [rule for everyone](https://reviews.llvm.org/H576) and for [beta testers](https://reviews.llvm.org/H511).
+  Note that right now there is no difference between beta and "normal" builds. 
+* the [merge_guards_bot user](https://reviews.llvm.org/p/merge_guards_bot/) account for writing comments.
+
+# Buildkite pipelines
+
+Buildkite allows [dynamically define pipelines as the output of a command](https://buildkite.com/docs/pipelines/defining-steps#dynamic-pipelines).
```
Comment (ChristianKuehnel) [inline docs/development.md:30]: add "to": allows **to** [dynamically

File: docs/development.md [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+export SRC=${BUILDKITE_BUILD_PATH}/llvm-premerge-checks
+rm -rf ${SRC}
+git clone --depth 1 --branch ${scripts_branch:-master} https://github.com/google/llvm-premerge-checks.git ${SRC}
+export SCRIPT_DIR=${SRC}/scripts
+${SCRIPT_DIR}/buildkite/build_branch_pipeline.py | tee /dev/tty | buildkite-agent pipeline upload
+```
+
+One typically edits corresponding script, not a pipeline definition in the Buildkite interface. [How to test changes](playbooks.md#testing-changes-before-merging).  
+
+# Life of pre-merge check
```
Comment (ChristianKuehnel) [inline docs/development.md:46]: add "a": Life of **a** pre-merge

File: docs/development.md [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+When new diff arrives for review it triggers Herald rule ("everyone" or "beta testers").
+
+That in sends an HTTP POST request to [**phab-proxy**](../phabricator-proxy) that submits a new buildkite job **diff-checks**.
```
Comment (ChristianKuehnel) [inline docs/development.md:50]: remove "in": That **in** sends

File: docs/development.md [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+All parameters from the original request are put in the build's environment with `ph_` prefix (to avoid shadowing any Buildkite environment variable).
+"scripts_branch" parameter defines which branch of llvm-premerge-checks to use.
+
+**diff-checks** pipeline ([create_branch_pipeline.py](../scripts/buildkite/create_branch_pipeline.py)) downloads a patch (or series of patches) and applies it to a fork of the llvm-project repository.
+Then it pushes a new state as a new branch (e.g. "phab-diff-288211") and triggers "premerge-checks" on it (all "ph_" env variables are passed to it).
```
Comment (ChristianKuehnel) [inline docs/development.md:55]: maybe add link where these branches can be found

**Review by ChristianKuehnel — APPROVED** [review]
LGTM, some minor wording hints. 

Switching user and developer view is a good idea! Thx for updating this!

File: README.md (hunk 1 above)
Comment (metaflow) [reply README.md]: done

Comment (metaflow) [reply README.md]: done

File: README.md (hunk 2 above)
Comment (metaflow) [reply README.md]: removed

File: README.md (hunk 3 above)
Comment (metaflow) [reply README.md]: done

File: README.md (hunk 4 above)
Comment (metaflow) [reply README.md]: sure

File: docs/development.md (hunk 5 above)
Comment (metaflow) [reply docs/development.md:10]: that's an online tool to generate TOC. Removed link

File: docs/development.md (hunk 10 above)
Comment (metaflow) [reply docs/development.md:62]: done

File: docs/playbooks.md (hunk 8 above)
Comment (metaflow) [reply docs/playbooks.md:15]: removed
